cumulus-client/src/main/java/mil/army/usace/hec/cumulus/client/controllers/CumulusDssFileController.java

```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    /**
+     * Download File to specified local file location.
+     *
+     * @param generatedDssFileDownloadData - Download object containing URL of file to download
+     * @param pathToLocalFile              - path in which file will be downloaded to
+     * @param listener                     - CumulusDssFileDownloadListener implementation that listens for download updates
+     * @return CompletableFuture           - Reference to async download process to be handled by caller
+     */
+    public CompletableFuture<Void> downloadToLocalFile(Download generatedDssFileDownloadData, Path pathToLocalFile,
+                                                       CumulusDssFileDownloadListener listener) throws CompletionException {
+        if (generatedDssFileDownloadData == null) {
+            throw new CompletionException(new IOException("Missing download Data"));
+        }
+        if (generatedDssFileDownloadData.getFile() == null) {
+            throw new CompletionException(new IOException("Missing DSS File URL for Download ID: " + generatedDssFileDownloadData.getId()));
+        }
+        return CompletableFuture.runAsync(() -> {
+            try {
+                CumulusFileDownloaderUtil.downloadFileToLocal(generatedDssFileDownloadData, pathToLocalFile, listener);
+            } catch (IOException e) {
+                throw new CompletionException(e);
+            }
+        });
+
+    }
+
+    private Download monitorAndRetrieveDownloadDataWhenComplete(ApiConnectionInfo apiConnectionInfo, Download initialDownloadData,
+                                                                CumulusDssFileGenerationListener listener) throws IOException {
+        DownloadsEndpointInput downloadsEndpointInput = new DownloadsEndpointInput(initialDownloadData.getId());
+        String progressIntervalVal = System.getProperty(PROGRESS_INTERVAL_PROPERTY_KEY);
+        if (progressIntervalVal == null) {
+            progressIntervalVal = "500";
+        }
+        int progressInterval = Integer.parseInt(progressIntervalVal);
+        long startTime = System.currentTimeMillis();
+        long progressStartTime = startTime;
+        long elapsedTimeSinceProgressChanged = 0L;
+        long maxStallTimeInMillis = TimeUnit.SECONDS.toMillis(MAX_ALLOWED_TIME_SECONDS);
+        Download downloadStatus = null;
+        int counter = 0;
+        String file = null;
+        int progress = 0;
+        while (file == null && elapsedTimeSinceProgressChanged <= maxStallTimeInMillis) {
+            counter++;
+            try {
+                downloadStatus = queryDownloadStatus(apiConnectionInfo, downloadsEndpointInput);
+                file = downloadStatus.getFile();
+                long totalElapsedTime = System.currentTimeMillis() - startTime;
+                if (listener != null) {
+                    listener.downloadStatusUpdated(downloadStatus);
```

Comment:
Why separate methods for a single event?

Reply:
Meaning: if I received three events, am I triggering UI updates for all three? Do I pick one? A single event would be more transparent.